Authentication in SvelteKit with Directus (#32)
directus-community/guest-authoring: sveltekit-authentication/index.md

---
title: 'Authentication in SvelteKit with Directus'

> description: 'How to authenticate users and integrate state managment with Cookies using SvelteKit SSR'

Reviewer: ```suggestion
description: 'Learn how to authenticate users and integrate state management with Cookies using SvelteKit with Server-Side Rendering (SSR).'
```

author:
  name: 'Eike Thies'
  avatar_file_name: 'avatar.jpg'
---

## Overview

> In this Guide we will show how to set up a complete authentication and authorization mechanism in SvelteKit using Directus as the Provider. The user session will be persisted via Server-Side Cookies and evidently be used within the Directus SDK to make authenticated Calls. As an example we will create a new Role showing how authorization works.

Reviewer: ```suggestion
In this guide, we will set up a complete authentication and authorization mechanism in SvelteKit using Directus as the Provider. The user session will be persisted via Server-Side Cookies and be used by the Directus SDK to make authenticated requests. As an example, we will create a new Role showing how authorization works.
```

> ## Primer Authentication
> 
> There are different ways how to store the user session. Here we will use the most secure form via Server-Side Cookies using the `Same-Site=Strict` attribute which will prevent CSRF Attacks. This means in order to be able to access the Directus Backend from the client directly, the Directus Backend will need to live on the same domain as the SvelteKit Frontend. For local development this will not matter. For more Information on this topic see https://andrewlock.net/understanding-samesite-cookies/

Reviewer: ```suggestion
## Understanding Server-Side Cookies

A user session is...

There are different ways to store a user session within web applications. We will use the most secure - server-side cookies using the `Same-Site=Strict` attribute which will prevent CSRF attacks. CSRF attacks are..

To access Directus from the client directly with this approach, Directus will need to live on the same domain as the SvelteKit application. For local development, this will not matter as... 

[Learn more about SameSite cookies](https://andrewlock.net/understanding-samesite-cookies/).
```

This is excellent context but needs a little but more - each paragraph ends with a prompt for you to complete.

Author: i extended the user session and CSRF paragraphs. not sure if we should go too deeply into this in the article. i would rather link some external explanations let me know if in the new version it is still not enough.

i removed the "same domain" paragraph as it was wrong. actually it will work in any case, as the cookie is only for the frontend and the token will get read from there and send to the directus backend if needed via the Directus SDK. that makes it work for any case (directus living in the cloud or deployed yourself on your domain)

## Adapt Directus Wrapper

> The following code examples will build up on the first Directus SvelteKit Blog Article. If you haven't seen it, you can check it out. Otherwise, with basic knowledge of SvelteKit you will be fine following along. First - let's change a few lines in our Directus SDK Wrapper, to utilize the Token and define a global Cookie Options Schema:

Reviewer: ```suggestion
In [Getting Started with Directus and SvelteKit](https://docs.directus.io/blog/getting-started-directus-sveltekit.html), we create a wrapper which makes the Directus SDK available to our project. In this project, we will adapt the wrapper to utilize the token and define a global cookie options schema:
```

```js [/libs/directus.js]
import { createDirectus, rest  } from "@directus/sdk"; // [!code --]
import { createDirectus, rest, authentication  } from "@directus/sdk"; // [!code ++]
import { PUBLIC_APIURL } from '$env/static/public'; // [!code --]
import { PUBLIC_APIURL,PUBLIC_COOKIE_DOMAIN } from '$env/static/public'; // [!code ++]

function getDirectusInstance(fetch,token) {

>   const options = fetch ? { globals: { fetch } } : {};

Reviewer: This doesn't need resolving by you but a note to self to run this through prettier before loading it in to be published. Spaces after commas is what caught my eye here.

  .with(authentication('cookie', { credentials: 'include' })) // [!code ++]
  .with(rest());

  if(token) directus.setToken(token); // [!code ++]

  return directus;
}
export default getDirectusInstance;

export const constructCookieOpts = (age) => { // [!code ++]
	return { // [!code ++]
			'domain': PUBLIC_COOKIE_DOMAIN, // [!code ++]
			// send cookie for every page // [!code ++]
			'path': '/', // [!code ++]
			// server side only cookie so you can't use `document.cookie` // [!code ++]
			'httpOnly': true, // [!code ++]
			// only requests from same site can send cookies // [!code ++]
			// https://developer.mozilla.org/en-US/docs/Glossary/CSRF // [!code ++]
			'sameSite': "strict", // [!code ++]
			// only sent over HTTPS in production // [!code ++]
			'secure': process.env.NODE_ENV === 'production', // [!code ++]
			// set cookie to expire after a given time // [!code ++]
			'maxAge': age // [!code ++]
		} // [!code ++]
	} // [!code ++]
```

Additionally we need to define the Domain the Cookie is valid for in the  `.env` file:

```
PUBLIC_APIURL= "https://directus.example.com"
PUBLIC_COOKIE_DOMAIN= "example.com" // [!code ++]

> ```

Reviewer: Flip the order - start by setting the variables before you use them.

> ## Setup Login/Signup Form

Reviewer: ```suggestion
## Create the Login Form
```

Let's start the user journey from adding a login/signup form:

```svelte [/signin/+page.svelte]
<script lang="ts">
	/** @type {import('./$types').PageData} */
	import { page } from '$app/stores';
  let email,password;
	const redirectedFrom = $page.url.searchParams.get('redirectedFrom')
</script>

<form
	action="?/login{redirectedFrom?`&redirectedFrom=${redirectedFrom}`:''}"
	method="POST">
	<div>
		<label for="email">Email</label>
		<input id="email" name="email" type="text" bind:value={email}	
		/>
	</div>

	<div>
		<label for="password">Password</label>
		<input id="password" name="password" type="password" required bind:value={password}	
		/>
	</div>

	<button type="submit">Log in</button>
	<button formaction="?/register{redirectedFrom?`&redirectedFrom=${redirectedFrom}`:''}">Register</button>
</form>
```

> The html template is straight forward. We only have a small function to let us redirect the user to a page he was trying to access before needing to login/signup.
> 
> The JavaScript form action below will handle the actual request.
> The Directus SDK is not saving HttpOnly Cookies, thus the Directus API is directly accessed from SvelteKit in order to be able to save the tokens in a secure cookie via SvelteKit's cookie handler.

Reviewer: ```suggestion
The template contains small function to let us redirect the user to a page they were trying to access before needing to login/signup. The JavaScript form action below will handle the request.
```

> The Directus SDK is not saving HttpOnly Cookies, thus the Directus API is directly accessed from SvelteKit in order to be able to save the tokens in a secure cookie via SvelteKit's cookie handler.

Can you rewrite this? Doesn't make sense to me as I read it.

Author: "The SvelteKit form action below will handle the actual request.
The Directus SDK is not saving Cookies with the HTTP-only Option enabled and we want to set the Cookie on the server, thus we need to access the Directus API directly from SvelteKit ourself and save the tokens in a secure cookie via SvelteKit's cookie handler."

better? i don't see a way around using SvelteKits cookie handler. sveltekit is in control here what is returned from the server, so the directus SDK is of no help here

```js [/signin/+page.server.js]
import { redirect,fail } from '@sveltejs/kit';
import { PUBLIC_APIURL } from '$env/static/public';
import { constructCookieOpts } from '$lib/directus';

> const REFRESH_TOKEN_EXPIRATION = 30; // in days - adapt this with your Directus Settings

Reviewer: What does the comment mean?

Author: i renamed the variable name and change the comment. it is refering to this setting. https://docs.directus.io/self-hosted/config-options.html#security it should just be in sync with it. for the directus cloud i guess it is 7 days, as stated in the docs. i changed it in the code accordingly to avoid any confusion.

	} catch (err) {
		return fail(400, {email,message:err.message});
	}

	redirect(302, redirectedFrom ? redirectedFrom : `/profile`)
}

const register = async ({ cookies, request, url }) => {
	const data = await request.formData();
	const email = data.get('email');
	const password = data.get('password');
	const redirectedFrom = url.searchParams.get('redirectedFrom')
	
  //First create the user
	let signupRequest = await fetch(`${PUBLIC_APIURL}/users`, {
		method: 'POST',
		headers: {
			'Content-Type': 'application/json',
			'user-agent':request.headers.get("user-agent")
		},
		body: JSON.stringify({
			email,
			password
		})
	});
	if(signupRequest.status >= 300){
		return fail(400, { email, message:await signupRequest.text() });
	}
	try {
		let tokens = await loginUser(request,email,password);

		cookies.set('access_token',tokens.access_token, constructCookieOpts(Math.floor(tokens.expires/1000)));
		cookies.set('refresh_token', tokens.refresh_token, constructCookieOpts(60 * 60 * 24 * REFRESH_TOKEN_EXPIRATION));
	} catch (err) {
		return fail(400, err.message);
	}
	
	redirect(302, redirectedFrom ? redirectedFrom : `/profile`)
}

export const actions = { login,register }
```

> 

Reviewer: ^ please break this down into smaller, explainable steps. This is a big snippet that contains too much functionality to just blindly copy and paste.

Author: i split it into two. more does not make sense, as it would create more overhead than help. i also added more comments.

## Setup Directus Roles

Before you open `http://localhost:5173/signup` and create a new test user, let us define a new role called `user` which all new users will inherit upon creating an account.

Within the Directus Admin App go to Settings -> Access Control -> Create new Role and name it `Users`. App Access should be enabled.

> ![A new role called User having App Access enabled.](new-role.webp)

Reviewer: Remove this - you describe it adequately and the image doesn't add value.

Also note down the role id from the url while you are in the Access Role page of the new Role.

To make this role the default for new registered users and also to enable public non-authenticated users to register themself, go to Access Control -> public and customize the create permission of the directus_users collection. Make sure that only email and password field permissions are set and as a default field use the following setting:
```
{
    "role": "<YOUR ROLE ID>"
}
```

> Here you will need to enter the Role ID from before. It could for example [look like this](default-role.webp).

Reviewer: This isn't valid image syntax. Please fix this

        let jwtPayload = cookies.get('access_token') ? jwt.decode(cookies.get('access_token')) : false;
  
        //check if token is expired and renew it if necessary
        if (isTokenExpired(jwtPayload) || !cookies.get('access_token')) {
          try {
            await refreshAccessToken(cookies);
            jwtPayload = cookies.get('access_token') ? jwt.decode(cookies.get('access_token')) : false;
          } catch (err) {
            cookies.delete('refresh_token', { path:'/' });
            cookies.delete('access_token', { path:'/' });
          }
        }
        
        event.locals.user = jwtPayload?.id;
        event.locals.token = cookies.get('access_token');
    }

    if (event.route.id && shouldProtectRoute(event.route.id) && !event.locals.user) {
        redirect(302,`/signin?redirectedFrom=${encodeURIComponent(url.pathname)}`)
    }

    // this is needed so that the response headers from SvelteKit do include the correct header to allow the browser to parse json requests
    return await resolve(event, {
        filterSerializedResponseHeaders: (key, value) => {
            return key.toLowerCase() === 'content-type'
        }
    });
}
```

> Every single request goes pass trough this file. It checks the tokens and if valid sets the user id and the token to the locals object, which can be accessed througout SvelteKits Load Functions. If the Access Token is expired, a new one will be generated. Finally, we can define routes under the `(protected)` Directory. The "protected" keyword will not appear in the url, but is only visible in our file structure.

Reviewer: ```suggestion
Every request will touch the logic in this file. It checks tokens and, if valid, sets the user id and the token to the locals object, which can be accessed throughout SvelteKit's load functions. If the Access Token is expired, a new one will be generated. Finally, we can define routes under the `(protected)` Directory. The "protected" keyword will not appear in the url, but is only visible in our file structure.
```

	return {
		user: locals.user,
		token: locals.token
	};
}
```

## Create Profile Page

The locals will now be used in every `+page.js` file to initialize the Directus SDK with the user's session token. To test this out, let's now create our `profile` page, which is only accessible if the user is logged in.

Create a file `/(protected)/profile/+page.js`

```js [/(protected)/profile/+page.js]
import getDirectusInstance from "$lib/directus";
import { readMe } from "@directus/sdk";
import { error } from "@sveltejs/kit";
export async function load({ parent, fetch }) {
  const { token } = await parent();
  const directus = getDirectusInstance(fetch, token);
  try {
    return {
      user: await directus.request(
        readMe({
          fields: ["*"],
        })
      ),
    };
  } catch (err) {
    error(404, "User not found");
  }
}

```

As you see, we are getting the token and initialize our Directus Instance as usual. This time however, we also give it the Access Token so that every request will now have the User's Session attached. In this case, we are reading the user's profile. If you try this without the token, the request will fail because of missing permissions.

> Let's continue writing the Html Template:

Reviewer: ```suggestion
To use data from Directus in the page we must export a `data` variable:
```

```svelte [/(protected)/profile/+page.svelte]
<script>
  export let data;
</script>
  
<p>This is only visible if you are logged in</p>

<p>Your Email: {data.user.email}</p>
```

You should now see a simple profile page if you are logged in under http://localhost:5173/profile

![A simple example of a profile page showing the users email](profile1.webp)

> ### Authenticated requests on the client and logging the user out

Reviewer: Split into two sections

Author: did you mean to use the `code group`markdown component so that the js and svelte file are shown side-by-side ? if not please explain

Author: sry i think i understand now, you mean logout as its own section. ok sure will do
